hare/metrics.go

```diff
@@ -45,4 +49,55 @@ var (
 		[]string{},
 		prometheus.ExponentialBuckets(4, 2, 3),
 	).WithLabelValues()
+
+	msgDelayTracker = newMessageDelayTracker()
 )
+
+func newMessageDelayTracker() *messageDelayTracker {
+	m := make(map[MessageType]map[bool]prometheus.Observer)
+	for _, t := range []MessageType{pre, status, proposal, commit, notify} {
+		m[t] = make(map[bool]prometheus.Observer)
+		for _, pos := range []bool{true, false} {
+			sign := "pos"
+			if !pos {
+				sign = "neg"
+			}
+
+			m[t][pos] = metrics.NewHistogramWithBuckets(
+				fmt.Sprintf("%v_%v_delay_seconds", strings.ToLower(t.String()), sign),
+				namespace,
+				fmt.Sprintf("delta between expected time of receipt of %v message and actual receipt time", t),
+				[]string{},
+				prometheus.ExponentialBuckets(0.2, 2, 10),
+			).WithLabelValues()
+		}
+	}
+	return &messageDelayTracker{
+		metrics: m,
+	}
+}
+
+// messageDelayTracker provides a convenient interface to track delays for hare
+// messages in prometheus histograms. The delay is calculated as the delta
+// between the start of a round and the time a message in that round was
+// received. There's a histogram for positive delays (i.e. the message arrived
+// after the start of the round) which would be the only expected values if all
+// nodes' clocks were synced. There's also a histogram for negative delays
+// (i.e. the message arrived before the start of the round) in this case nodes'
+// clocks must be out of sync. The gnerated metric names look like
+// notify_neg_delay_seconds or notify_pos_delay_seconds with the first segment
+// differing based on the message type.
+type messageDelayTracker struct {
+	metrics map[MessageType]map[bool]prometheus.Observer
+}
+
+func (m messageDelayTracker) trackDelay(t MessageType, clock RoundClock) {
+	subMap, ok := m.metrics[t]
+	if !ok {
+		// Untracked message type
+		return
+	}
+	seconds := time.Since(clock.RoundEnd(t.round() - 1)).Seconds()
```

Comment:
there is a difference btwn message type value and round number.

for example. status round (val 0) can have round number of 0/4/8/12. the param to `clock.RoundEnd()` expect the actual round number, which is recorded in `msg.Round()`

Reply:
i think the current implementation works correctly only for the first iteration (which is a bug)

Reply:
Are you referring to the round clock or my use of it here?

Reply:
Ok, figuring out a fix.